on-demand download

problame opened the pull request:
This is not yet ready for review. Only for myself to run tests in CI.

problame commented:
Wanted to rebase this to run the CI, that then prompted a detour that resulted in https://github.com/neondatabase/neon/pull/3026

I'll rebase this branch onto that PR now.

problame commented:
TODOs (discussed with @hlinnaka ):

- [ ]  endpoint to on-demand download all "active" layers in the layer map, i.e., all layers needed to reconstruct at the latest LSN
- [ ] Make GC work without downloading => add test
- [ ] Separate metric for physical size on page server vs S3 (PS: `resident_physical_size` , `s3_physical_size`)

hlinnaka commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -1088,21 +1100,18 @@ impl Timeline {
                         continue;
```
Fix "Download them now" comment above the loop

hlinnaka commented on pageserver/src/tenant/size.rs:
```diff
@@ -364,12 +362,9 @@ async fn calculate_logical_size(
-        TimelineAtLsnSizeResult(timeline, lsn, size_res)
-    })
-    .await
+    let _permit = permit;
+    let size_res = timeline.calculate_logical_size(lsn).await;
+    Ok(TimelineAtLsnSizeResult(timeline, lsn, size_res))
```
If `calculate_logical_size`  takes a long time, without having to download anything, this can now hog the async thread for a very long time. It will finish eventually though. I guess that's OK for now, it's a more general problem we have with our model where we freely perform std synchronous I/O in async code.

problame replied:
I only changed this because `timeline.calculate_logical_size` **had** to become `async` because it now downloads layers internally.
Is there a way to tell Tokio that a future will be compute-heavy? I'd use it for this.

problame replied:
Cool, will implement that.

problame replied:
Ok, I thought this was straight-forward, and it is.
Then I looked at other callers of `Timeline::calculate_logical_size` and found that in `try_spawn_size_init_task` I removed the `spawn_blocking`  as well.
So, presumably, we have the same problem there.

So, will duplicate @koivunej 's snippet from above to there as well.

problame replied:
Hm actually that's the only other caller, so, can move that into `Timeline::calculate_logical_size`.

problame replied:
Oh no, can't share the code because `size.rs` has this semaphore :see_no_evil:

hlinnaka commented on test_runner/regress/test_tenant_relocation.py:
```diff
@@ -461,13 +461,16 @@ def test_tenant_relocation(
         # is no longer involved, and if it is, we will see the error
         pageserver_http.tenant_detach(tenant_id)
-
+        assert_tenant_status(pageserver_http, tenant_id, "Stopping")
```
never mind, this was changed in a later commit

problame commented:
Status update:

We have two sorts of regress-test errors that are caused by changes in this branch:

---

```
test_pageserver_metrics_removed_after_detach

AssertionError: assert {'pageserver_..._seconds_sum'} == set()
  Extra items in the left set:
  'pageserver_io_operations_seconds_bucket'
  'pageserver_io_operations_seconds_count'
  'pageserver_io_operations_seconds_sum'
  Full diff:
  - set(,
  - )
  + {
  +  'pageserver_io_operations_seconds_bucket',
  +  'pageserver_io_operations_seconds_count',
  +  'pageserver_io_operations_seconds_sum',
  + }
```

This one is fully reproducible and I'm half-way through understanding it.
What I know is that something calls `VirtualFile::metdata()` from `DeltaLayerWriter::finish()` or `ImageLayerWriter::finish()` after we've already dropped `TimelineMetrics.
I added `VirtualFile::metadata()` in one of the latest commits in this branch to support the new `current_logical_size` semantics.
My gut feeling is that the flush loop writes out a last layer when it's shut down, and somehow that outlives the timeline object.
Or it's some other flush loop.
Need to investigate more here.

---

```
test_ondemand_download.py

test_runner/regress/test_ondemand_download.py:356: in test_download_remote_layers_api
    assert attached_size < filled_size
E   assert 29818880 < 29818880
```

and similar errors related to size estimation.

I think this is due to my recent change `WIP make GC work for remote layers`.
This adds one more retry_get, and I think that causes all layers to be downloaded in some cases.

We decided that it's OK to ship this PR even if it effectively re-downloads all layer.
So, if we want to meet the deadline, we can just throw out all those size assertions.

Still hoping to make some progress there, though.

problame commented:
TODOs:

- [ ] Ensure we retry failing downloads for at least 30 seconds.
  - Test using failpoints.

problame commented:
Rebased & squashed the changs that I made on top of Heikki's initial on-demand download commit.
Let's see what the test run brings.

The remaining issue that I'm aware of is the (lack of) retries by retry_get / retry_get_with_timeout.

problame commented:
- [ ] download_remote_layer task should be sensitive to task_mgr::shutdown

problame commented:
Force-push: rebase + following commits + Cherry-pick of @SomeoneToIgnore 's style nits:

```
commit 58f2fbac817fa6089596e2e1bea606bc3f051fae                                               
Author: Christian Schwarz <christian@neon.tech>                                               
Date:   Tue Dec 20 15:05:56 2022 +0100                                                        
                                                                                              
    remove retry_get_with_download + rename retry_get to with_ondemand_download + doc comment 
```

```
commit f19231be4aefb6aafc84e5cb0878500468c2575d                           
Author: Christian Schwarz <christian@neon.tech>                           
Date:   Tue Dec 20 16:24:22 2022 +0100                                    
                                                                          
    replace RemoteLayer's DownloadState logic with a Sempahore            
                                                                          
    The previous implementation was quite verbose and had a race          
    condition - we would need an async condvar to make it work correctly. 
                                                                          
    As a replacement, use a tokio::Semaphore that we close if the         
    download was successful.                                              
    We use a similar, albeit not identical approach, for initial logical  
    size computation.                                                     
```


I think this is now in shape with what we discussed in the meeting.

I didn't add an extra test because the `test_download_remote_layers_api` test already ensures that you can call `Timeline::download_remote_layer` multiple times to retry.

Some eyes on `replace RemoteLayer's DownloadState logic with a Sempahore` would be appreciated.

SomeoneToIgnore commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -2819,14 +2902,339 @@ impl Timeline {
+        );
+
+        let initial_info = DownloadRemoteLayersTaskInfo {
+            task_id: format!("{task_id}"),
```
Curious, what was wrong with the original `PageserverTaskId` that it cannot be left in the `DownloadRemoteLayersTaskInfo` as is? This way, we could make the latter `Copy` and avoid a few `clone`'s around.

problame replied:
It's not `serde::Serialize`

SomeoneToIgnore replied:
Yes, but we can move it into the api module and make such?

problame replied:
Not sure if we want to leak those details. In particular because task numbers are sequential, and re-used after restart. I'd much rather prefer random UUIDs that will never be re-used.
But didn't want to add a UUID to each task, so, went for using strings that we can change to UUIDs later on.

SomeoneToIgnore replied:
> leak those details

Technically, they are leaked already, just in an inconvenient form.

But replacing it with the UUID later makes more sense indeed.

SomeoneToIgnore reviewed (approved):
Pushed a few nits, but overall, surprisingly (and uncanny, as if some error is missed) polished compared to what I've seen a few months ago.

I see a few things up for potential follow-ups, that were not mentioned recently before:
* `PageReconstructError` interface is odd, as discussed on the call, it's also quite error-prone due to ability to be turned into random `anyhow::Error` via `?` almost enywhere
I think, there should be `PageReconstructResult<T>` instead, with a few utility methods for ignoring the download error.
* `impl PersistentLayer for RemoteLayer` and following downcasts are super odd, we should better define what is a historical layer and avoid using traits for such definition.
* I think we should try and use the new client with errors introduced in https://github.com/neondatabase/neon/pull/3134 for testing instead of failpoints

hlinnaka commented on pageserver/src/http/openapi_spec.yml:
```diff
@@ -82,11 +82,6 @@ paths:
         schema:
           type: string
           description: Controls calculation of current_logical_size_non_incremental
-      - name: include-non-incremental-physical-size
```
Is this used from the console? I guess not, or the end-to-end tests would fail. Todo: remove it from the cloud repo's autogenerated files.

problame replied:
https://github.com/neondatabase/cloud/pull/3358

hlinnaka commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -158,7 +161,7 @@ pub struct Timeline {
     // List of child timelines and their branch points. This is needed to avoid
     // garbage collecting data that is still needed by the child timelines.
-    pub gc_info: RwLock<GcInfo>,
+    pub gc_info: tokio::sync::RwLock<GcInfo>,
```
Why this change? This lock protects just a few variables..

Hmm, I can see the held while making pretty complicated calls in `update_gc_info`. `update_gc_info` calls `find_lsn_for_timestamp` in particular. I think we should refactor `update_gc_info` so that it would do the `find_lsn_for_timestamp` call before acquiring this lock.

(not a blocker, I'm ok to punt that to a followup PR)

problame replied:
Good find, it's not really necessary. Will amend the original preliminary refactor commit that introduced it in the next force push.

hlinnaka commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -1026,11 +1067,13 @@ impl Timeline {
-            // We'll download a fresh copy of the layer file below.
+            // remote index file?
+            // If so, rename_to_backup those files & replace their local layer with
+            // a RemoteLayer in the laye rmap so that we re-download them on-demand.
```
```suggestion
            // a RemoteLayer in the layer map so that we re-download them on-demand.
```

hlinnaka commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -1590,6 +1634,18 @@ impl Timeline {
                 //info!("CHECKING for {} at {} on historic layer {}", key, cont_lsn, layer.filename().display());
 
+                //
+                // Is this a remote layer? If so, create a Future that the caller can use to
```
Outdated comment: we don't create a Future here anymore

hlinnaka commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -2819,14 +2902,339 @@ impl Timeline {
+    ///
+    /// It's safe to call this function for the same layer concurrently.
+    ///
+    /// On success, returns the layer's size.
```
actually returns ()

hlinnaka commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -2819,14 +2902,339 @@ impl Timeline {
+    /// Don't use the success value for any acounting purposes, though, since
+    /// every concurrent caller gets that value (=> you would double-account).
+    ///
+    /// If there is a download error, the function retries indefinitely, using an exponential back-off.
```
no retries here

hlinnaka commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -2819,14 +2902,339 @@ impl Timeline {
+    /// every concurrent caller gets that value (=> you would double-account).
+    ///
+    /// If there is a download error, the function retries indefinitely, using an exponential back-off.
+    /// If you need a timeout, simply stop polling: we're cancellation-safe because the download happens in a spearate task_mgr task.
```
If you just stop polling, the download continues in the background right? Because it's spawned to a separate task,

```suggestion
    /// If you need a timeout, simply stop polling: we're cancellation-safe because the download happens in a seeparate task_mgr task.
```

hlinnaka commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -2819,14 +2902,339 @@ impl Timeline {
+
+                // In case we failed and there are other waiters, this will make one
+                // of them retry the download in a new task.
+                // XXX: This resets the exponential backoff because it's a new call to
```
no retries

problame replied:
What do you mean by this? once we release the permit, another waiter will come through the semaphore and create a new task with a new `download_layer_file` call.

hlinnaka replied:
Never mind. I just saw the "exponential backoff" and though this is an obsolete comment because there is no retry logic. The comment explains it, I just didn't read it fully.

problame commented:
- [ ] Test failure `test_tenant_relocation[without_load-minor]` is reproducible for me. Will spend some time tonight debugging this, but I'm not familiar with the test, so if someone else is, I'd appreciate help there.

SomeoneToIgnore commented:
Release Allure report

https://neon-github-public-dev.s3.amazonaws.com/reports/pr-3013/release/3742243072/index.html#suites/9d2450a537238135fd4007859e09aca7/ea0db6f759ec2111/

contains `repo/new_pageserver/pageserver.log` 

https://neon-github-public-dev.s3.amazonaws.com/reports/pr-3013/release/3742243072/data/attachments/c37a223d71fcfe40.log

with 

```
2022-12-20T20:02:47.674778Z ERROR error importing base backup between 0/16FA398 and 0/16FA398: Failed to import basebackup

Caused by:
    invalid digit found in string

Stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.66/src/backtrace/libunwind.rs:93:5
      backtrace::backtrace::trace_unsynchronized
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.66/src/backtrace/mod.rs:66:5
      backtrace::backtrace::trace
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.66/src/backtrace/mod.rs:53:14
   1: anyhow::backtrace::capture::Backtrace::create
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/anyhow-1.0.66/src/backtrace.rs:216:13
      anyhow::backtrace::capture::Backtrace::capture
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/anyhow-1.0.66/src/backtrace.rs:204:17
   2: anyhow::error::<impl core::convert::From<E> for anyhow::Error>::from
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/anyhow-1.0.66/src/error.rs:547:25
   3: <core::result::Result<T,F> as core::ops::try_trait::FromResidual<core::result::Result<core::convert::Infallible,E>>>::from_residual
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/result.rs:2105:27
      pageserver::import_datadir::import_slru
             at /Users/someonetoignore/work/neon/neon/pageserver/src/import_datadir.rs:198:17
   4: pageserver::import_datadir::import_file
             at /Users/someonetoignore/work/neon/neon/pageserver/src/import_datadir.rs:515:9
   5: pageserver::import_datadir::import_basebackup_from_tar
             at /Users/someonetoignore/work/neon/neon/pageserver/src/import_datadir.rs:326:36
   6: pageserver::tenant::UninitializedTimeline::import_basebackup_from_tar::{{closure}}::{{closure}}
             at /Users/someonetoignore/work/neon/neon/pageserver/src/tenant.rs:249:13
      tokio::runtime::enter::exit
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/enter.rs:91:9
   7: tokio::runtime::scheduler::multi_thread::worker::block_in_place
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/scheduler/multi_thread/worker.rs:339:9
   8: tokio::task::blocking::block_in_place
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/task/blocking.rs:77:9
      pageserver::tenant::UninitializedTimeline::import_basebackup_from_tar::{{closure}}
             at /Users/someonetoignore/work/neon/neon/pageserver/src/tenant.rs:248:9
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/future/mod.rs:91:19
   9: pageserver::page_service::PageServerHandler::handle_import_basebackup::{{closure}}::{{closure}}
             at /Users/someonetoignore/work/neon/neon/pageserver/src/page_service.rs:402:70
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/future/mod.rs:91:19
  10: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-0.1.37/src/instrument.rs:272:9
  11: pageserver::page_service::PageServerHandler::handle_import_basebackup::{{closure}}
             at /Users/someonetoignore/work/neon/neon/pageserver/src/page_service.rs:367:5
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/future/mod.rs:91:19
  12: <pageserver::page_service::PageServerHandler as utils::postgres_backend_async::Handler>::process_query::{{closure}}
             at /Users/someonetoignore/work/neon/neon/pageserver/src/page_service.rs:856:18
  13: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/future/mod.rs:91:19
  14: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/future/future.rs:124:9
  15: utils::postgres_backend_async::PostgresBackend::process_message::{{closure}}
             at /Users/someonetoignore/work/neon/neon/libs/utils/src/postgres_backend_async.rs:419:74
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/future/mod.rs:91:19
      utils::postgres_backend_async::PostgresBackend::run_message_loop::{{closure}}
             at /Users/someonetoignore/work/neon/neon/libs/utils/src/postgres_backend_async.rs:268:79
  16: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/future/mod.rs:91:19
      utils::postgres_backend_async::PostgresBackend::run::{{closure}}
             at /Users/someonetoignore/work/neon/neon/libs/utils/src/postgres_backend_async.rs:203:67
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/future/mod.rs:91:19
      pageserver::page_service::page_service_conn_main::{{closure}}
             at /Users/someonetoignore/work/neon/neon/pageserver/src/page_service.rs:194:60
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/future/mod.rs:91:19
  17: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::future::future::Future>::poll
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/panic/unwind_safe.rs:296:9
      <futures_util::future::future::catch_unwind::CatchUnwind<Fut> as core::future::future::Future>::poll::{{closure}}
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.25/src/future/future/catch_unwind.rs:36:42
      <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/panic/unwind_safe.rs:271:9
      std::panicking::try::do_call
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panicking.rs:492:40
      std::panicking::try
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panicking.rs:456:19
      std::panic::catch_unwind
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panic.rs:137:14
      <futures_util::future::future::catch_unwind::CatchUnwind<Fut> as core::future::future::Future>::poll
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.25/src/future/future/catch_unwind.rs:36:9
  18: <tokio::task::task_local::TaskLocalFuture<T,F> as core::future::future::Future>::poll::{{closure}}
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/task/task_local.rs:348:35
      tokio::task::task_local::LocalKey<T>::scope_inner
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/task/task_local.rs:233:19
      <tokio::task::task_local::TaskLocalFuture<T,F> as core::future::future::Future>::poll
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/task/task_local.rs:345:13
      <tokio::task::task_local::TaskLocalFuture<T,F> as core::future::future::Future>::poll::{{closure}}
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/task/task_local.rs:348:35
      tokio::task::task_local::LocalKey<T>::scope_inner
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/task/task_local.rs:233:19
      <tokio::task::task_local::TaskLocalFuture<T,F> as core::future::future::Future>::poll
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/task/task_local.rs:345:13
  19: pageserver::task_mgr::task_wrapper::{{closure}}
             at /Users/someonetoignore/work/neon/neon/pageserver/src/task_mgr.rs:320:10
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/future/mod.rs:91:19
  20: tokio::runtime::task::core::CoreStage<T>::poll::{{closure}}
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/task/core.rs:184:17
      tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/loom/std/unsafe_cell.rs:14:9
      tokio::runtime::task::core::CoreStage<T>::poll
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/task/core.rs:174:13
      tokio::runtime::task::harness::poll_future::{{closure}}
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/task/harness.rs:480:19
      <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/panic/unwind_safe.rs:271:9
  21: std::panicking::try::do_call
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panicking.rs:492:40
      std::panicking::try
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panicking.rs:456:19
      std::panic::catch_unwind
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panic.rs:137:14
      tokio::runtime::task::harness::poll_future
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/task/harness.rs:468:18
      tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/task/harness.rs:104:27
      tokio::runtime::task::harness::Harness<T,S>::poll
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/task/harness.rs:57:15
  22: tokio::runtime::task::raw::poll
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/task/raw.rs:194:5
  23: tokio::runtime::task::raw::RawTask::poll
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/task/raw.rs:134:18
      tokio::runtime::task::LocalNotified<S>::run
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/task/mod.rs:385:9
      tokio::runtime::scheduler::multi_thread::worker::Context::run_task::{{closure}}
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/scheduler/multi_thread/worker.rs:421:13
      tokio::coop::with_budget::{{closure}}
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/coop.rs:102:9
      std::thread::local::LocalKey<T>::try_with
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/thread/local.rs:445:16
      std::thread::local::LocalKey<T>::with
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/thread/local.rs:421:9
  24: tokio::coop::with_budget
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/coop.rs:95:5
      tokio::coop::budget
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/coop.rs:72:5
      tokio::runtime::scheduler::multi_thread::worker::Context::run_task
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/scheduler/multi_thread/worker.rs:420:9
  25: tokio::runtime::scheduler::multi_thread::worker::Context::run
  26: tokio::runtime::scheduler::multi_thread::worker::run::{{closure}}
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/scheduler/multi_thread/worker.rs:372:17
      tokio::macros::scoped_tls::ScopedKey<T>::set
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/macros/scoped_tls.rs:61:9
  27: tokio::runtime::scheduler::multi_thread::worker::run
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/scheduler/multi_thread/worker.rs:369:5
  28: tokio::runtime::scheduler::multi_thread::worker::Launch::launch::{{closure}}
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/scheduler/multi_thread/worker.rs:348:45
      <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/blocking/task.rs:42:21
  29: tokio::runtime::task::core::CoreStage<T>::poll::{{closure}}
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/task/core.rs:184:17
      tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/loom/std/unsafe_cell.rs:14:9
      tokio::runtime::task::core::CoreStage<T>::poll
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/task/core.rs:174:13
      tokio::runtime::task::harness::poll_future::{{closure}}
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/task/harness.rs:480:19
      <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/panic/unwind_safe.rs:271:9
      std::panicking::try::do_call
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panicking.rs:492:40
      std::panicking::try
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panicking.rs:456:19
      std::panic::catch_unwind
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panic.rs:137:14
      tokio::runtime::task::harness::poll_future
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/task/harness.rs:468:18
      tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/task/harness.rs:104:27
      tokio::runtime::task::harness::Harness<T,S>::poll
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/task/harness.rs:57:15
  30: tokio::runtime::task::raw::RawTask::poll
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/task/raw.rs:134:18
      tokio::runtime::task::UnownedTask<S>::run
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/task/mod.rs:422:9
      tokio::runtime::blocking::pool::Task::run
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/blocking/pool.rs:111:9
      tokio::runtime::blocking::pool::Inner::run
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/blocking/pool.rs:346:17
  31: tokio::runtime::blocking::pool::Spawner::spawn_thread::{{closure}}
             at /Users/someonetoignore/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/blocking/pool.rs:321:13
      std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/sys_common/backtrace.rs:122:18
  32: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/thread/mod.rs:501:17
      <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/panic/unwind_safe.rs:271:9
      std::panicking::try::do_call
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panicking.rs:492:40
      std::panicking::try
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panicking.rs:456:19
      std::panic::catch_unwind
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panic.rs:137:14
      std::thread::Builder::spawn_unchecked_::{{closure}}
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/thread/mod.rs:500:30
      core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/ops/function.rs:248:5
  33: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/alloc/src/boxed.rs:1872:9
      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/alloc/src/boxed.rs:1872:9
      std::sys::unix::thread::Thread::new::thread_start
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/sys/unix/thread.rs:108:17
  34: __pthread_deallocate
```

looking further.

SomeoneToIgnore commented:
Looks like my paths error is mac-specific and not related, not sure about the remote one though.
Yet, on Linux machine, I see locally some other errors:

```
2022-12-20T20:25:21.127793Z  WARN download RemotePath("tenants/74ee8b079a0e437eb0afea7d26a07209/timelines/c6f8b99fd67664546d3ed2136562bf51/000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__00000000016965B0-0000000001696629") failed, will retry (attempt 8): Failed to open a download stream for layer with remote storage path 'RemotePath("tenants/74ee8b079a0e437eb0afea7d26a07209/timelines/c6f8b99fd67664546d3ed2136562bf51/000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__00000000016965B0-0000000001696629")': No file found for the remote object id given

and then a lot of repeating errors
```

looks like we should cancel downloads on remotely deleted files.
The good way would be to do so via the task manager, but as a fix, looks like we can react on the corresponding error differently?

hlinnaka commented:
> Test failure test_tenant_relocation[without_load-minor] is reproducible for me. Will spend some time tonight debugging this, but I'm not familiar with the test, so if someone else is, I'd appreciate help there.

Garden-variety bug in `get_reconstruct_data: it returns incorrect timeline in the PageReconstructError::NeedDownload it creates, if it had to recurse to the ancestor timeline. This fixes it:
```
diff --git a/pageserver/src/tenant/timeline.rs b/pageserver/src/tenant/timeline.rs
index 9a104ec42..cae6aeafb 100644
--- a/pageserver/src/tenant/timeline.rs
+++ b/pageserver/src/tenant/timeline.rs
@@ -1641,7 +1641,7 @@ impl Timeline {
                 if let Some(remote_layer) = super::storage_layer::downcast_remote_layer(&layer) {
                     info!("need remote layer {}", layer.traversal_id());
                     return Err(PageReconstructError::NeedDownload(
-                        Weak::clone(&self.myself),
+                        Weak::clone(&timeline.myself),
                         Arc::downgrade(&remote_layer),
                     ));
                 }
```

problame commented on pageserver/src/http/openapi_spec.yml:
```diff
@@ -266,6 +256,55 @@ paths:
+        "202":
+          description: |
+            Task scheduled successfully.
+            The returned token identifies the operation.
```
More generally, I'm wondering whether we should document this API endpoint at all.
It's not supposed to be used by the console anyways, just a failsafe for manual intervention during an incident.

SomeoneToIgnore replied:
We could create yet another spec nearby, but feels bulletproof to exclude from console's codegen whatever is not planned to be used there.

problame commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -1590,6 +1634,18 @@ impl Timeline {
+                // download the missing layer, and return it.
+                //
+                if let Some(remote_layer) = super::storage_layer::downcast_remote_layer(&layer) {
+                    info!("need remote layer {}", layer.traversal_id());
```
I've seen log lines in the relocation test where this message has no surrounding span. So, grepping for tenant_id won't show this message.

problame commented:
Force push:

- Addressed https://github.com/neondatabase/neon/pull/3013#discussion_r1053559974 (amended initial refactoring commit)
- Rebase on top of latest `main`

Only significant rebase conflict was with @lubennikovaav 's 4235f97c6a3276d90a1c3630fa78d4b1495df26 (metrics collection). Resolved as planned in https://docs.google.com/document/d/12AFpvKY-7FZdR5a4CaD6Ir_rI3QokdCLSPJ6upHxJBo/edit# by using the resident_size metrics instead of the (removed) `get_physical_size`.

- [ ] @lubennikovaav , please double-check the change in `pageserver/src/billing_metrics.rs`  (including following the function call).

problame commented:
Note that `test_metric_collection` fails about 50% of the time due to

```
AssertionError: assert not ['2022-12-20T22:24:05.642541Z ERROR wal_connection_manager{tenant=5ba5b3597edb4a7c69ca23640ea6fbe8 timeline=ce9e2646e8...=ce9e2646e8213c0a88dbf1ff45cbf440}: wal receiver task finished with an error: Failed to open walreceiver connection\n']
```

Checked the logs, I don't think it's this PR's fault.

problame commented:
- [ ] need to revisit `get_rel_exists` call sites and check for ignored `PageReconstructError`

```
https://neon-github-public-dev.s3.amazonaws.com/reports/pr-3013/release/3742243072/data/attachments/e0b8d6c6afc8b95a.log
022-12-20T16:38:04.421950Z ERROR wal_connection_manager{tenant=74ee8b079a0e437eb0afea7d26a07209 timeline=51e580a513365545d3785eff00a2d665}: wal receiver task finished with an error: walreceiver connection handling failure

Caused by:
    0: could not ingest record at {lsn}
    1: layer file needs to be downloaded

Stack backtrace:
   0: pageserver::walingest::WalIngest::handle_rel_extend
   1: pageserver::walingest::WalIngest::put_rel_page_image
             at /__w/neon/neon/pageserver/src/walingest.rs:927:9
      pageserver::walingest::WalIngest::ingest_decoded_block
             at /__w/neon/neon/pageserver/src/walingest.rs:348:13
      pageserver::walingest::WalIngest::ingest_record
             at /__w/neon/neon/pageserver/src/walingest.rs:282:13
   2: pageserver::walreceiver::walreceiver_connection::handle_walreceiver_connection::{{closure}}
             at /__w/neon/neon/pageserver/src/walreceiver/walreceiver_connection.rs:251:25
```

problame commented on test_runner/regress/test_remote_storage.py:
```diff
@@ -134,21 +133,31 @@ def test_remote_storage_backup_and_restore(
 
     # assert cannot attach timeline that is scheduled for download
-    # FIXME implement layer download retries
+    # FIXME: we used to keep retrying the initial download on failure.
```
Outdated since Heikkis commit

hlinnaka replied:
Hmm, yeah I think so.

problame replied:
Good, will clean up this test wrt on-demand download.

hlinnaka replied:
This test and the comments have become quite messy. I'll take a stab at cleaning it up

problame replied:
I guess we raced around this, just pushed my cleanup change and then noticed yours.

problame commented on test_runner/regress/test_remote_storage.py:
```diff
@@ -134,21 +133,31 @@ def test_remote_storage_backup_and_restore(
     tenant_status = client.tenant_status(tenant_id)
     log.info("Tenant status with active failpoint: %s", tenant_status)
-    # FIXME implement layer download retries
+    # FIXME see above
```
See above.

problame commented on test_runner/fixtures/neon_fixtures.py:
```diff
@@ -1839,6 +1937,9 @@ def __init__(self, env: NeonEnv, port: PageserverPort, config_override: Optional
             ".*query handler for 'pagestream.*failed: Timeline .* is not active",  # timeline delete in progress
+            # FIXME: on-demand download causes these, need to triage
+            ".*could not compact, repartitioning keyspace failed: layer file needs to be downloaded",
+            ".*Gc failed, retrying in .*: layer file needs to be downloaded",
```
Do these log messages still apply?
@SomeoneToIgnore changed the log message text for `NeedsDownload` recently.

Can't find them in the codebase.

Remove these from `allowed_errors`

problame replied:
Removed them.

lubennikovaav commented on pageserver/src/billing_metrics.rs:
```diff
@@ -186,7 +186,7 @@ pub async fn collect_metrics_task(
             ));
 
-            let timeline_size = timeline.get_physical_size();
+            let timeline_size = timeline.get_resident_physical_size();
```
Maybe we should rename the metric to `ResidentPhysicalSize` as well to be consistent? WDYT?

It is not used anywhere yet, so we can do it without breaking compatibility.

problame replied:
Yes, will do that.

problame commented:
Force push: rebase on top of latest `main`. Intentional conflict with openapi_spec.yml because of https://github.com/neondatabase/neon/pull/3168 

I'm not aware of any more open TODOs.

If there are no more test failures, I'll pretty up the commit history and do one more force push, then merge.

problame commented:
Test failure `test_metric_collection[noop]`: obviously flaky.

Console failure: no idea how to debug these. Re-running it before squashing.

problame commented:
Console re-run succeeded. Squashing now.

lubennikovaav commented:
> Test failure `test_metric_collection[noop]`: obviously flaky.
> 

My bad. Fixed in https://github.com/neondatabase/neon/pull/3173
